src/Compilers/CSharp/Portable/Parser/LanguageParser.cs

```diff
@@ -6373,29 +6380,33 @@ private bool IsTerm()
 
         private TypeSyntax ParseTypeCore(
             bool parentIsParameter,
-            bool isOrAs,
+            bool isPattern,
+            bool rightOfAs,
```

Comment:
It looks like these two parameters are mutually exclusive. Consider adding an assert, or better use one enum parameter instead.